### ISSUE TITLE
Backstage doesn't recognize branches that contain forward slashes

### ISSUE THREAD

**Issue by taras** [backstage/backstage#2815]
Backstage relies on [git-url-parse](https://github.com/IonicaBizau/git-url-parse) to parse URLs. This library has a [bug](https://github.com/IonicaBizau/git-url-parse/issues/104) that causes it to incorrectly parse URLs that point to branches with forward slashes in them. 

## Expected Behavior

I expect to be able to use branches with `/` because it's a useful way to organized branches into groups. 

## Current Behavior

If I attempt to create a template from a branch that uses a `/` in its name, the operation will fail because Backstage (via git-url-parse) will assume that I'm trying to get code from branch called `tm`. It will ignore everything past first `/`.

## Possible Solution

🤷‍♀️

## Steps to Reproduce

1. Try to create a template from https://github.com/spotify/backstage/blob/dependabot/npm_and_yarn/testing-library/jest-dom-5.11.4/plugins/scaffolder-backend/sample-templates/create-react-app/template.yaml

## Context

Some of the projects I'm involved in require that each team member prefixes their branch with their initials. It's a habit at this point to start every branch name with `tm/`.

**Comment by freben** [comment]
Oof - thanks for reporting, of course this should work and I always do the same.

That library has been a bit less than stellar, I wonder if there are alternatives. Then again, just looking at the url in your example, it looks hard to deduce where the ref ends and the path begins...

**Comment by taras** [comment]
I created a [runkit](https://runkit.com/taras/5f80311544ac68001b8ebdf1) that has every GitHub URL parsing library that I could find. None of them support properly detecting the branch name. We might need a completely different strategy for doing this.

**Comment by Rugvip** [comment]
https://github.com/spotify/backstage/blob/blob/blob/blob/blob/blob/blob.txt

There is actually a bit of method to this madness though. Git itself has a limitation that you can only create branches that can be represented in the filesystem. For example, you can't have both the branch `a/b` and `a/b/c`, because when you create `a/b` `b` is a text file on disk, and therefore you can't create `a/b/c` since `b` would have to be a folder for that.

With that in mind it means there's at least a deterministic way to interpret the URL, the annoying bit is that it includes using the `GitHub` API to discover the branches that are present in the repo.

**Comment by taras** [comment]
> the annoying bit is that it includes using the GitHub API to discover the branches that are present in the repo

I was thinking the same thing.

There could be another approach that could work but I don't know how universal it wold be. 

GitHub website doesn't seem to care about extra forward slashes. We could for example use `//` forward slash to represent beginning and end of a ref. It would also be annoying, but could be better than auto detecting because the user is expressing their intention by adding the extra forward slash. 

For example, 'https://github.com/spotify/backstage/blob//dependabot/npm_and_yarn/testing-library/jest-dom-5.11.4//package.json' still leads to the same file on the GitHub website.

**Comment by Rugvip** [comment]
@taras tbh we want users to be able to just copy the address from the address bar. Otherwise we could just go with our own shorter format too

**Comment by taras** [comment]
true true. I'm going to email a friend at Microsoft to see if he can connect us with someone at GitHub. I want to find out from them if there is a way that they'd recommend.

**Comment by IonicaBizau** [comment]
Like I commented in https://github.com/IonicaBizau/git-url-parse/pull/114#issuecomment-706091677, I don't think there is a way to split the url itself into branch name and filename... Pinging @izuzak here just in case he has any ideas how this can be improved. :sparkle:

**Comment by timbonicus** [comment]
@mtlewis and I spent a hot second looking at the [code in question](https://github.com/backstage/backstage/blob/9eccc589b7ebe3c2675a29f650bf2417d89d2573/packages/backend-common/src/reading/GithubUrlReader.ts#L267). As mentioned, there's not enough information in the URL to tell branch name versus path when the branch has a slash. I didn't find any alternative URL formats that work with GitHub where this might be possible.

This code could have some additional logic, though. I think we could either:

1. List all branches for the repo to match a substring
2. Catch the 404 and retry, expanding the branch name to the next `/`

Since (1) could suffer on a repo with a ton of branches, and could have false positives, the second seems preferable to me. We could retry N times, expanding to the next slash each time, where N is a fairly low number.

**Comment by freben** [comment]
One issue iirc is that there's a lot of code that just uses git-url-parse or things like [resolveUrl](https://github.com/backstage/backstage/blob/master/packages/integration/src/types.ts#L50) that aren't even async in the first place so there's very little room for doing smart or retrying or remote-querying things. Feels to me like it unfortunately would take a big refactoring to make it good.

**Comment by eperdeme** [comment]
Evening, 
I'm not a developer but doing loops to attempt to work out if its a branch or a path seems inelegant. 

In other projects I've seen the URL is normally constructed for git repos with parameters that get extracted to let the client libs know what the branch is;

For example;

Terraform
``
git::https://example.com/vpc.git?ref=v1.2.0
``
Helmfile
``
- name: polaris
  url: git+https://github.com/repo/test@deploy/helm?ref=master
``

kustomize
``
github.com/kubernetes-sigs/kustomize/examples/multibases?ref=v1.0.6
``

**Comment by freben** [comment]
@eperper thanks for those pointers - interesting. One thing to take into account here is convenience of copy+paste as mentioned by @Rugvip above. Having to form a custom url might be error prone. That being said, I think that approach can be worth considering nonetheless, especially if it's an "alternative form" that you can use only when necessary or if it's easily available somehow in the web interface besides the address bar of the browser.

**Comment by FilipPyrek** [comment]
Hi guys, what's a status on this one? Unfortunately it's a TechDocs show stopper for us.. 😢

**Comment by freben** [comment]
Sorry but this is still a current issue. There's some discussion in the comments above. Having unencoded slashes in the URL makes it super problematic to parse and understand GitHub URLs - there's no surefire way to actually know where the branch ends and the path begins, without resorting to hammering the GitHub API with queries to resolve the situation - and this needs to be done in a lot of places where the code isn't even async ready, but rather expects to be able to do that parsing instantly and locally.

**Comment by FilipPyrek** [comment]
@freben I like the idea of "alternative form" which would solve the problem but didn't require any groundbreaking changes. I like format where `ref` is supplied as a query string. That solution would make it extensible also for other information in future - if needed. 
Do you guys at @backstage prefer to implement this yourself? Or is this okay to be done by a community? Do you have any requirements & tips how you would like this to be implemented (and where)?

**Comment by freben** [comment]
@FilipPyrek Yeah that's probably the most viable way forward that we have seen so far, at least. We could make sure to implement support for the alternative form _as well as_ the original native form - we do want to keep the convenience of copy+paste from the browser.

There's TONS of direct usages of `git-url-parse` across the code base. If we went for this strategy, we would have to rewrite them all to use some wrapper function e.g. in `@backstage/integration` instead. There's also probably a fair bit of code that does direct string manipulation, or does `new URL(str).pathname.split('/')` and similar trickery. So yeah, there will be a bunch of chasing down code that isn't up to par with the new requirement.

**Comment by freben** [comment]
And to answer your question, we'd appreciate help with it.

**Comment by FilipPyrek** [comment]
Understood and totally agree @freben.

I will see what available capacity we will have and if somebody from us will start working on that, I will let you guys know. 👍

**Comment by dallasgutauckis** [comment]
De-stale-inating. This is still a bug. We had someone else run into it again today.

**Comment by dhhuynh2** [comment]
Also ran into this recently while utilizing the [validation endpoint](https://github.com/backstage/backstage/blob/23d032e987e8b65c2b17a6837199f7a69d217c8c/plugins/catalog-backend/src/service/createRouter.ts#L241). Feeding branches with forward slashes broke the validator.

**Comment by abogdanov37** [comment]
Hi everyone. I faced with the same problem for GitLab. And I offer the foloowing solution
1. Add to integration config option to use only root `catalog-info.yaml` placement
2. If that option is enabled we parse all path  before the `catalog-info.yaml` as branch name
3. If not we have current behavior

For example we have following  gitlab url `https://gitlabcom/group/subgroup/project/-/blob/branch_a/branch_b/catalog-info.yaml`
we need to fix that line to get in file path only file name
https://github.com/backstage/backstage/blob/30314172ed18b3c10f49a14a88d35270b69041c2/packages/integration/src/gitlab/core.ts#L80

**Comment by abogdanov37** [comment]
Hi guys! I work on solution suggested me above. This solution is so bad ((( I implement another one solution 
Main idea is check by existing branches in repository. When we create url to get source we
1/ get branch name as is in current code
2/ get all branches from repo which names start with name got on step 1
3/ match path to getted branch names 
I test this solution in my company and it works
I add code below. 
I need review for that idea. 
If it's ok I add this changes in integration backend

```
// Converts
// from: https://gitlab.com/groupA/teams/teamA/subgroupA/repoA/-/blob/branch/filepath
// to:   https://gitlab.com/api/v4/projects/projectId/repository/files/filepath?ref=branch
export async function buildProjectUrl(
    target: string,
    projectID: Number,
    config: GitLabIntegrationConfig,
): Promise<URL> {
    try {
        const url = new URL(target);

        const branchAndFilePath = url.pathname
            .split('/blob/')
            .slice(1)
            .join('/blob/');

        let fileName: string;
        let branch: string;
        const [br, ...filePath] = branchAndFilePath.split('/');
        let repoBranches = await getProjectBranches(target, "^" + br, projectID, config);
        let findedBranch = checkExistingBranches(repoBranches, branchAndFilePath);
        if (findedBranch != undefined) {
            let path = branchAndFilePath.replace(findedBranch, "");
            fileName = path.split('/').filter(p => p).join("/");
            branch = findedBranch;
        } else {
            branch = br;
            fileName = filePath.join('/');
        }
        
        const relativePath = getGitLabIntegrationRelativePath(config);

        url.pathname = [
            ...(relativePath ? [relativePath] : []),
            'api/v4/projects',
            projectID,
            'repository/files',
            encodeURIComponent(decodeURIComponent(fileName)),
            'raw',
        ].join('/');

        url.search = `?ref=${branch}`;

        return url;
    } catch (e) {
        throw new Error(`Incorrect url: ${target}, ${e}`);
    }
}

// Get all branches started with some-name
export async function getProjectBranches(
    target: string,
    branchSearchTerm: string,
    projectID: Number,
    config: GitLabIntegrationConfig,
): Promise<Array<string>> {
    const url = new URL(target);

    if (!url.pathname.includes('/blob/')) {
        throw new Error('Please provide full path to yaml file from GitLab');
    }

    try {
        // Get gitlab relative path
        const relativePath = getGitLabIntegrationRelativePath(config);

        // Convert
        // to: https://gitlab.com/api/v4/projects/groupA%2Fteams%2FsubgroupA%2FteamA%2Frepo
        const repoIDLookup = new URL(
            `${url.origin}${relativePath}/api/v4/projects/${projectID}/repository/branches?search=${encodeURIComponent(
                branchSearchTerm,
            )}`,
        );

        const response = await fetch(
            repoIDLookup.toString(),
            getGitLabRequestOptions(config),
        );

        const data = await response.json();

        if (!response.ok) {
            throw new Error(
                `GitLab Error '${data.error}', ${data.error_description}`,
            );
        }
        let branches = data as Array<any>;
        return branches.map(b => b.name);
    } catch (e) {
        throw new Error(`Could not get GitLab project ID for: ${target}, ${e}`);
    }
}
```

**Comment by abogdanov37** [comment]
> Hi guys! I work on solution suggested me above. This solution is so bad ((( I implement another one solution Main idea is check by existing branches in repository. When we create url to get source we 1/ get branch name as is in current code 2/ get all branches from repo which names start with name got on step 1 3/ match path to getted branch names I test this solution in my company and it works I add code below. I need review for that idea. If it's ok I add this changes in integration backend
> 
> ```
> // Converts
> // from: https://gitlab.com/groupA/teams/teamA/subgroupA/repoA/-/blob/branch/filepath
> // to:   https://gitlab.com/api/v4/projects/projectId/repository/files/filepath?ref=branch
> export async function buildProjectUrl(
>     target: string,
>     projectID: Number,
>     config: GitLabIntegrationConfig,
> ): Promise<URL> {
>     try {
>         const url = new URL(target);
> 
>         const branchAndFilePath = url.pathname
>             .split('/blob/')
>             .slice(1)
>             .join('/blob/');
> 
>         let fileName: string;
>         let branch: string;
>         const [br, ...filePath] = branchAndFilePath.split('/');
>         let repoBranches = await getProjectBranches(target, "^" + br, projectID, config);
>         let findedBranch = checkExistingBranches(repoBranches, branchAndFilePath);
>         if (findedBranch != undefined) {
>             let path = branchAndFilePath.replace(findedBranch, "");
>             fileName = path.split('/').filter(p => p).join("/");
>             branch = findedBranch;
>         } else {
>             branch = br;
>             fileName = filePath.join('/');
>         }
>         
>         const relativePath = getGitLabIntegrationRelativePath(config);
> 
>         url.pathname = [
>             ...(relativePath ? [relativePath] : []),
>             'api/v4/projects',
>             projectID,
>             'repository/files',
>             encodeURIComponent(decodeURIComponent(fileName)),
>             'raw',
>         ].join('/');
> 
>         url.search = `?ref=${branch}`;
> 
>         return url;
>     } catch (e) {
>         throw new Error(`Incorrect url: ${target}, ${e}`);
>     }
> }
> 
> // Get all branches started with some-name
> export async function getProjectBranches(
>     target: string,
>     branchSearchTerm: string,
>     projectID: Number,
>     config: GitLabIntegrationConfig,
> ): Promise<Array<string>> {
>     const url = new URL(target);
> 
>     if (!url.pathname.includes('/blob/')) {
>         throw new Error('Please provide full path to yaml file from GitLab');
>     }
> 
>     try {
>         // Get gitlab relative path
>         const relativePath = getGitLabIntegrationRelativePath(config);
> 
>         // Convert
>         // to: https://gitlab.com/api/v4/projects/groupA%2Fteams%2FsubgroupA%2FteamA%2Frepo
>         const repoIDLookup = new URL(
>             `${url.origin}${relativePath}/api/v4/projects/${projectID}/repository/branches?search=${encodeURIComponent(
>                 branchSearchTerm,
>             )}`,
>         );
> 
>         const response = await fetch(
>             repoIDLookup.toString(),
>             getGitLabRequestOptions(config),
>         );
> 
>         const data = await response.json();
> 
>         if (!response.ok) {
>             throw new Error(
>                 `GitLab Error '${data.error}', ${data.error_description}`,
>             );
>         }
>         let branches = data as Array<any>;
>         return branches.map(b => b.name);
>     } catch (e) {
>         throw new Error(`Could not get GitLab project ID for: ${target}, ${e}`);
>     }
> }
> ```

@ferben @Rugvip guys need a feedback

**Comment by Rugvip** [comment]
@abogdanov37 to be honest I think we want to avoid any solution that adds an extra request to the URL reading. This problem seems to still cause less trouble than rate limits, which we would start hitting a lot sooner if we add in another request. That's overall why this is such a tricky problem, there's no solution that doesn't add an extra request that we know of.

We discussed this a bit more today, and we're thinking that the best approach to this would be to parse the URL during registration in the catalog, and along with that have our own internal URL format that is able to carry the information. That's also a good opportunity to standardize URL parsing and handling across all different SCM integrations. It's a significant chunk of work though, and unless someone want to pick that up and drive it all the way, it's not something I feel is worth prioritizing over other work at this point.

**Comment by abogdanov37** [comment]
> @abogdanov37 to be honest I think we want to avoid any solution that adds an extra request to the URL reading. This problem seems to still cause less trouble than rate limits, which we would start hitting a lot sooner if we add in another request. That's overall why this is such a tricky problem, there's no solution that doesn't add an extra request that we know of.
> 
> We discussed this a bit more today, and we're thinking that the best approach to this would be to parse the URL during registration in the catalog, and along with that have our own internal URL format that is able to carry the information. That's also a good opportunity to standardize URL parsing and handling across all different SCM integrations. It's a significant chunk of work though, and unless someone want to pick that up and drive it all the way, it's not something I feel is worth prioritizing over other work at this point.

Hi @Rugvip! Thanks for a feedback! I understand your concern. Idea with internal url format seems good but we have no time to to wait solution and have no resources to do it (((  
We can implement the same behavior without addition request bu we should have in config list of branches containing `/`. As for me it's not so good because we should manage this list by hand. But may be it helps you some how. 
For next step I try to create and public package for Backstage with my solution. I hope it helps some one.

**Comment by santunioni** [comment]
I am very interested in resolving this issue. What do you think of this proposal? https://github.com/backstage/backstage/pull/16407

**Comment by freben** [comment]
@santunioni Sorry that we (I) have been slow - love your idea, will get to reviewing it properly eventually

**Comment by Rugvip** [comment]
Way forward with this issue is what was started in https://github.com/backstage/backstage/pull/16407 but unfortunately abandoned

**Comment by VladMasarik** [comment]
yello, I started looking at the abandoned PR, and would like to continue from there.

So far I think identified that issue to be here
https://github.com/backstage/backstage/blob/master/packages/integration/src/helpers.ts#L98

Specifically, we pass in `url="../a.yaml"`, and the base path is changed, but the query should be actually changed. 
```
Expected: "https://github.com/backstage/backstage/tree/branchName/withSlashes?path=a.yaml#L17"
Received: "https://github.com/backstage/backstage/tree/a.yaml?path=test/README.md#L17"
```
I would like to edit this part, but do you think this is the correct place? Or would you actually know of a better / proffered place for that fix?

**Comment by freben** [comment]
@VladMasarik That's the _default_ resolver, and each individual integration may or may not choose to use it. Each integration class' `resolveUrl` method is what's actually being consumed, and for example [the Azure one](https://github.com/backstage/backstage/blob/5949408aaaa795044caa9f96fdc1273b62b0c099/packages/integration/src/azure/AzureIntegration.ts#L57) does something completely different and more complex. So this will have to be resolved on a per-provider level since they all have different idiosyncrasies. In the case of your example URLs, it'd be [here](https://github.com/backstage/backstage/blob/5949408aaaa795044caa9f96fdc1273b62b0c099/packages/integration/src/github/GithubIntegration.ts#L54).

Now, to make things worse, URLs are used in a huge number of places, and they are acted on in various ways. Not just `resolveUrl`, but for example parsing them and taking out their component parts. A lot of code uses the `git-url-parse` library directly for this since the `integrations` package doesn't offer this yet. See [here](https://github.com/backstage/backstage/blob/5949408aaaa795044caa9f96fdc1273b62b0c099/plugins/catalog-import/src/api/GitHub.ts#L29) for example. We need to make sure that these various loose usages of URLs also work with the secondary format that has the query parameter. If you can double check that it'd be great.

GIven these circumstances I think we'll have a hard time supporting the URL format that has the path inside the URL path itself rather than in a `path` query parameter. We just can't know for sure where the branch name ends and the path starts...

**Comment by VladMasarik** [comment]
@freben I am still assuming that for the Default and Github resolver it could be useful, and it wouldn't be perfect, but at least something.

However, do I understand correctly that although we might resolve the issue of parsing branches with `/` in them in one place, it won't fix all of the issues because in other places the library cannot parse the `/`? As in, lets say we would actually fix the issue for github resolver, but when manipulating the same URL in other place everything would suddenly stop working?

I looked at the library directly, there is an open issue as well, but it has been seemingly abandoned because it is tricky for the same issue you mentioned. If I were still interested in fixing this issue, do you think fixing the library first would be preferred?

Also, I am getting the impression that you would very much welcome this fix, however you don't expect it to be actually implemented because of the mentioned issues, and as such you are not very excited to draw this out even further? :D 

Just as an extra context, I did not read all the messages that were discussed in the last two years that this PR was open. My main reason for trying to revive this is the last comment 
> Way forward with this issue is what was started in https://github.com/backstage/backstage/pull/16407 but unfortunately abandoned

**Comment by freben** [comment]
Yeah I think the core of my worry is: This looks like a point fix in a single place but there's probably a lot more places that still won't work with URLs on the proposed form. We need help validating that things _actually work_ throughout Backstage with URLs on a new form and it might be a bit of an undertaking. What I need is a bit of confidence in merging whatever gets proposed. Because if we "officially" say that "yeah if your branch contains slashes, just register the URL on this form instead" and that makes e.g. the scaffolder not work at all for those URLs ... then we've introduced new problems :) Does that make sense? On the surface I see a proposed change in the resolve method that looks like it makes sense in isolation, but I don't know the impact that the change has on a larger scale, and have limited time to dig through and validate things by hand on my own end.

**Comment by VladMasarik** [comment]
> What I need is a bit of confidence in merging whatever gets proposed

Ahh yeah, that is understandable. In that case, I would like to first take care of the issue in the `git-url-parse` library first, as it seems that there might not even be a simple resolution to it. Since it turned out to be a bit larger undertaking, I won't promise any progress (as a note for anyone else reading this), but I will try to add any new findings in here. Glad I first asked before resurrecting the old PR right away :D 

Thanks for sharing your perspective Frederik :ok_hand: 

Just some references for the PRs / issues I was talking about in the `git-url-parse` library
https://github.com/IonicaBizau/git-url-parse/pull/114
https://github.com/IonicaBizau/git-url-parse/issues/104

**Comment by piatkiewicz** [comment]
Maybe a message could be a little more descriptive. I can't reproduce it now for new repo and branch when creating some simple catalog-info, but in some old project I have this message when trying to import catalog-info.yaml from branch with a slash in the name
![image](https://github.com/backstage/backstage/assets/8793897/1d6503c8-1c50-4748-8a6f-b6ea84313680)
For branches without a slash it imports fine.

**Comment by VladMasarik** [comment]
ping, still a problem AFAIK. the last time i was looking at this, I think it was possible to fix, however when I was searching for the package, I found only a few matches so i was confused because someone mentioned it is used in a lot of places. However, later I found out that I was looking at the newest package, and not an older version, and then I found like 50 places where it was used, and never got to fixing all of them.

**Comment by henrik-koren** [comment]
This is still an issues

**Comment by glimchb** [comment]
ping, still a problem for us

**Comment by joecombopiano** [comment]
FYI It looks like `git-url-parse` published a fix for this[ yesterday (August 22, 2024) .](https://github.com/IonicaBizau/git-url-parse/releases/tag/15.0.0)

**Comment by freben** [comment]
Hmm only if one passes in all known branch names, which we don't. And parsing generally happens in a non-async context which is part of the reason that we're in this bind - if we weren't, we could have asked the remote service directly instead of leaning on string parsing. I may be mistaken but I'm afraid this particular change can't be leveraged 🤔

**Comment by abdulshere** [comment]
Hi @freben, Usage of "/" for the branch name is a common valid practice in the industry and not supporting this will be a great setback for us to use some of the backstage plugins, we would highly appreciate if you could leverage the recent fix on git-url-parse to support the additional "ref" parameter which is optional.

**Comment by henrik-koren** [comment]
> published

What is a valid solution to address this issues in away that it acceptable  @freben

**Comment by nimmi-exxcellent** [comment]
I don't think that there will ever be a perfect solution without going async and fetching available branches from remote.
The question is if it makes sense to implement a workaround that works for most people.

One idea would be to assume a certain list of common branch prefixes like `feat/` or `fix/` and try to parse the url with those in mind. Because assuming is generally a bad idea it would probably be better to default to an empty list of prefixes and make them configurable in the backstage config.

This approach would solve this issue for most people from what i have gathered. It won't work with branch names containing slashes that are not prefixes, like `fix/branch/name`. The feature could be extended of course to include postfixes as well or even regular expressions but that seems a bit much.

Unfortunately this won't address dynamic/unknown pre/postfixes like originally mentioned in this issue.

**Comment by mixa1901** [comment]
We want to contribute to fix the issue. @freben, could you please  hint which direction would be acceptable for you?

**Comment by freben** [comment]
That doesn't sound like it would work in practice. The only real way forward is probably to make the integrations package have an async URL parser that can do things like this on demand.

**Comment by andrew-mclachlan** [comment]
We are working around this issue with a permalink but, as you can tell, it would need to be updated should the catalog-info.yaml need to change.

**Comment by boroamir** [comment]
Hi all, It appears the github-url-parse issue was fixed. Is there any downside to upgrading the npm package to 15.0.0?

**Comment by freben** [comment]
It wasn't fixed in a sustainable way.

The core of the problem is that there are huge numbers of places in the code that are NOT in an async context that parse URLs synchronously. And due to the format of GitHub URLs, there is absolutely no way to reliably parse them without making a request to the GitHub API to help parse them. I don't think that any amount of change to github-url-parse itself can get around this.

Example: `https://github.com/backstage/backstage/blob/foo/foo/foo` could be a reference to the file `foo` in the branch `foo/foo`, or a reference to the file `foo/foo` in the branch `foo`.

A thorough solution probably requires us to invent a core integrations API/service that has the ability to do this in a smart way, asynchronously if needed - and then to rewrite all parts of the project that aren't in an async context to be able to call this.

**Comment by enri-kapaj** [comment]
Is there any activity done on this issue  or any work around to make it possible to upload a file from a feature branch in format feature/<branch_name>

**Comment by franfrandev** [comment]
Working on a personal project but stumbled into exactly this problem. It seems impossible to fix this synchronously as noted in the issue. One workaround would be to make the parsing function "best effort", where it would return a list of possible combinations instead of only one. Then, the asynchronous code with networking capabilities may try them all until finding a satisfying branch & path, much like: https://github.com/backstage/backstage/issues/2815#issuecomment-876740188

<details>
  <summary>Something along those lines (Rust implementation)</summary>

```rust
#[derive(Debug)]
pub struct CandidatesIter {
    base_url: String,
    pin: String,
    rest: String,
    exhausted: bool,
}

impl CandidatesIter {
    fn new(base_url: String, pin: String, rest: String) -> Self {
        Self {
            base_url,
            pin,
            rest,
            exhausted: false,
        }
    }
}

#[derive(Debug)]
#[cfg_attr(test, derive(PartialEq))]
pub struct CommitPin {
    tag_or_commit: Option<String>,
    path: String,
}

impl Iterator for CandidatesIter {
    type Item = CommitPin;

    fn next(&mut self) -> Option<Self::Item> {
        if self.exhausted {
            return None;
        }

        let item = CommitPin {
            tag_or_commit: Some(self.pin.clone()),
            path: self.rest.clone(),
        };

        if !self.rest.is_empty() {
            let (part, rest) = match self.rest.split_once('/') {
                Some((part, rest)) => (part, rest.to_string()),
                None => (self.rest.as_str(), "".to_string()),
            };
            self.pin += &('/'.to_string() + part);
            self.rest = rest.to_string();
        } else {
            self.exhausted = true;
        }

        Some(item)
    }
}

#[derive(Debug)]
pub enum RepoLink {
    BaseUrl(String),
    Candidates(CandidatesIter),
}

impl RepoLink {
    pub fn base_url(&self) -> &str {
        match self {
            RepoLink::BaseUrl(base_url) => base_url,
            RepoLink::Candidates(candidates) => &candidates.base_url,
        }
    }
}

/// Incrementally attempt to build a longer branch with a shorter path with the `Iterator` primitive in the `RepoLink::Candidates` variant.
/// Returns a `RepoLink::BaseUrl` if there is no ambiguity.
pub fn repo_url_to_candidates(url: &str) -> Option<RepoLink> {
    let Ok(base_url) = normalize_repo_url(url) else {
        return None;
    };

    let Some(stripped_url) = url.strip_prefix(&base_url) else {
        return None; // invalid url format
    };
    if stripped_url.chars().next() == Some('/') {
        let stripped_url = &stripped_url[1..];
        let Some((kind, rest)) = stripped_url.split_once('/') else {
            return Some(RepoLink::BaseUrl(base_url.to_string()));
        };

        if !(kind == "tree" || kind == "releases" || kind == "blob") {
            return Some(RepoLink::BaseUrl(base_url.to_string()));
        }

        let (pin, rest) = match rest.split_once('/') {
            Some((pin, rest)) => (pin.to_string(), rest.to_string()),
            None => (rest.to_string(), "".to_string()),
        };

        Some(RepoLink::Candidates(CandidatesIter::new(
            base_url, pin, rest,
        )))
    } else {
        Some(RepoLink::BaseUrl(base_url.to_string()))
    }
}
```
</details>
